Fix deadlock when stopping playback through json

a1rwulf opened the pull request:
## Description
If you start playback and stop it fast (having refreshrate switching enabled) there is a high chance of a deadlock.
The fist approach was to make the resolution set asynchronous, which seems to be wrong.

See stacktrace:

```
libc.so.6!syscall() (/build/glibc-OTsEL5/glibc-2.27/sysdeps/unix/sysv/linux/x86_64/syscall.S:38)
Main Thread:
libstdc++.so.6!std::__atomic_futex_unsigned_base::_M_futex_wait_until(unsigned int*, unsigned int, bool, std::chrono::duration<long, std::ratio<1l, 1l> >, std::chrono::duration<long, std::ratio<1l, 1000000000l> >) (Unknown Source:0)
std::__atomic_futex_unsigned<2147483648u>::_M_load_and_test_until(std::__atomic_futex_unsigned<2147483648> * const this, unsigned int __assumed, unsigned int __operand, bool __equal, std::memory_order __mo, bool __has_timeout, std::chrono::seconds __s, std::chrono::nanoseconds __ns) (/usr/include/c++/7/bits/atomic_futex.h:102)
std::__atomic_futex_unsigned<2147483648u>::_M_load_and_test_until_impl<std::chrono::duration<long, std::ratio<1l, 1000000000l> > >(std::__atomic_futex_unsigned<2147483648> * const this, unsigned int __assumed, unsigned int __operand, bool __equal, std::memory_order __mo, const std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1, 1000000000> > > & __atime) (/usr/include/c++/7/bits/atomic_futex.h:140)
std::__atomic_futex_unsigned<2147483648u>::_M_load_when_equal_until<std::chrono::duration<long, std::ratio<1l, 1000000000l> > >(std::memory_order __mo, unsigned int __val, std::__atomic_futex_unsigned<2147483648> * const this) (/usr/include/c++/7/bits/atomic_futex.h:199)
std::__atomic_futex_unsigned<2147483648u>::_M_load_when_equal_for<long, std::ratio<1l, 1000l> >(const std::chrono::duration<long, std::ratio<1, 1000> > & __rtime, std::memory_order __mo, unsigned int __val, std::__atomic_futex_unsigned<2147483648> * const this) (/usr/include/c++/7/bits/atomic_futex.h:172)
std::__future_base::_State_baseV2::wait_for<long, std::ratio<1l, 1000l> >(std::__future_base::_State_baseV2 * const this, const std::chrono::duration<long, std::ratio<1, 1000> > & __rel) (/usr/include/c++/7/future:351)
std::__basic_future<bool>::wait_for<long, std::ratio<1l, 1000l> >(const std::__basic_future<bool> * const this, const std::chrono::duration<long, std::ratio<1, 1000> > & __rel) (/usr/include/c++/7/future:700)
CThread::Join(CThread * const this, unsigned int milliseconds) (/home/a1rwulf/devel/kodi/xbmc/threads/Thread.cpp:253)
CThread::StopThread(CThread * const this, bool bWait) (/home/a1rwulf/devel/kodi/xbmc/threads/Thread.cpp:204)
CVideoPlayer::CloseFile(CVideoPlayer * const this, bool reopen) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:714)
CApplicationPlayer::CloseFile(CApplicationPlayer * const this, bool reopen) (/home/a1rwulf/devel/kodi/xbmc/ApplicationPlayer.cpp:56)
CApplicationPlayer::ClosePlayer(CApplicationPlayer * const this) (/home/a1rwulf/devel/kodi/xbmc/ApplicationPlayer.cpp:39)
CApplication::StopPlaying(CApplication * const this) (/home/a1rwulf/devel/kodi/xbmc/Application.cpp:3309)
PLAYLIST::CPlayListPlayer::OnApplicationMessage(PLAYLIST::CPlayListPlayer * const this, KODI::MESSAGING::ThreadMessage * pMsg) (/home/a1rwulf/devel/kodi/xbmc/PlayListPlayer.cpp:963)
KODI::MESSAGING::CApplicationMessenger::ProcessMessage(KODI::MESSAGING::CApplicationMessenger * const this, KODI::MESSAGING::ThreadMessage * pMsg) (/home/a1rwulf/devel/kodi/xbmc/messaging/ApplicationMessenger.cpp:242)
KODI::MESSAGING::CApplicationMessenger::ProcessMessages(KODI::MESSAGING::CApplicationMessenger * const this) (/home/a1rwulf/devel/kodi/xbmc/messaging/ApplicationMessenger.cpp:215)
CApplication::Process(CApplication * const this) (/home/a1rwulf/devel/kodi/xbmc/Application.cpp:4046)
CXBApplicationEx::Run(CXBApplicationEx * const this, const CAppParamParser & params) (/home/a1rwulf/devel/kodi/xbmc/XBApplicationEx.cpp:64)
XBMC_Run(bool renderGUI, const CAppParamParser & params) (/home/a1rwulf/devel/kodi/xbmc/platform/xbmc.cpp:72)
```

```
libpthread.so.0!futex_wait_cancelable(unsigned int expected, unsigned int * futex_word) (/build/glibc-OTsEL5/glibc-2.27/sysdeps/unix/sysv/linux/futex-internal.h:88)
libpthread.so.0!__pthread_cond_wait_common(pthread_mutex_t * mutex, pthread_cond_t * cond) (/build/glibc-OTsEL5/glibc-2.27/nptl/pthread_cond_wait.c:502)
libpthread.so.0!__pthread_cond_wait(pthread_cond_t * cond, pthread_mutex_t * mutex) (/build/glibc-OTsEL5/glibc-2.27/nptl/pthread_cond_wait.c:655)
libstdc++.so.6!std::condition_variable::wait(std::unique_lock<std::mutex>&) (Unknown Source:0)
std::_V2::condition_variable_any::wait<XbmcThreads::CRecursiveMutex>(std::_V2::condition_variable_any * const this, XbmcThreads::CRecursiveMutex & __lock) (/usr/include/c++/7/condition_variable:251)
XbmcThreads::ConditionVariable::wait(XbmcThreads::ConditionVariable * const this, CCriticalSection & lock) (/home/a1rwulf/devel/kodi/xbmc/threads/Condition.h:38)
XbmcThreads::TightConditionVariable<bool volatile&>::wait<CCriticalSection>(XbmcThreads::TightConditionVariable<bool volatile&> * const this, CCriticalSection & lock) (/home/a1rwulf/devel/kodi/xbmc/threads/Condition.h:83)
CEvent::Wait(CEvent * const this) (/home/a1rwulf/devel/kodi/xbmc/threads/Event.h:87)
KODI::MESSAGING::CApplicationMessenger::SendMsg(KODI::MESSAGING::CApplicationMessenger * const this, KODI::MESSAGING::ThreadMessage && message, bool wait) (/home/a1rwulf/devel/kodi/xbmc/messaging/ApplicationMessenger.cpp:146)
KODI::MESSAGING::CApplicationMessenger::SendMsg(KODI::MESSAGING::CApplicationMessenger * const this, uint32_t messageId, int param1, int param2, void * payload) (/home/a1rwulf/devel/kodi/xbmc/messaging/ApplicationMessenger.cpp:161)
CGraphicContext::SetVideoResolution(CGraphicContext * const this, RESOLUTION res, bool forceUpdate) (/home/a1rwulf/devel/kodi/xbmc/windowing/GraphicContext.cpp:390)
CVideoPlayer::OpenVideoStream(CVideoPlayer * const this, CDVDStreamInfo & hint, bool reset) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:3580)
CVideoPlayer::OpenStream(CVideoPlayer * const this, CCurrentStream & current, int64_t demuxerId, int iStream, int source, bool reset) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:3447)
CVideoPlayer::OpenDefaultStreams(CVideoPlayer * const this, bool reset) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:882)
CVideoPlayer::Prepare(CVideoPlayer * const this) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:1231)
CVideoPlayer::Process(CVideoPlayer * const this) (/home/a1rwulf/devel/kodi/xbmc/cores/VideoPlayer/VideoPlayer.cpp:1307)
CThread::Action(CThread * const this) (/home/a1rwulf/devel/kodi/xbmc/threads/Thread.cpp:282)
CThread::<lambda(CThread*, std::promise<bool>)>::operator()(CThread *, std::promise<bool>) const(const CThread::<lambda(CThread*, std::promise<bool>)> * const __closure, CThread * pThread, std::promise<bool> promise) (/home/a1rwulf/devel/kodi/xbmc/threads/Thread.cpp:140)
std::__invoke_impl<void, CThread::Create(bool)::<lambda(CThread*, std::promise<bool>)>, CThread*, std::promise<bool> >(std::__invoke_other, CThread::<lambda(CThread*, std::promise<bool>)> &&, CThread *&&, std::promise<bool> &&)(CThread::<lambda(CThread*, std::promise<bool>)> && __f,  __args#0,  __args#1) (/usr/include/c++/7/bits/invoke.h:60)
std::__invoke<CThread::Create(bool)::<lambda(CThread*, std::promise<bool>)>, CThread*, std::promise<bool> >(CThread::<lambda(CThread*, std::promise<bool>)> &&, CThread *&&, std::promise<bool> &&)(CThread::<lambda(CThread*, std::promise<bool>)> && __fn,  __args#0,  __args#1) (/usr/include/c++/7/bits/invoke.h:95)
```



## How Has This Been Tested?
I've used following script to execute the deadlock case.
After the change it does not happen again.

```
#!/bin/bash
while true
do
    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Open","sender":"xyz","id":1564573142676,"params":{"item":{"channelid":1}}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2
    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Stop","sender":"xyz","id":1564573142430,"params":{"playerid":1}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2

    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Open","sender":"xyz","id":1564573142676,"params":{"item":{"channelid":2}}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2
    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Stop","sender":"xyz","id":1564573142430,"params":{"playerid":1}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2

    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Open","sender":"xyz","id":1564573142676,"params":{"item":{"channelid":3}}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2
    curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.Stop","sender":"xyz","id":1564573142430,"params":{"playerid":1}}' -H 'content-type: application/json;' http://$1:8080/jsonrpc
    sleep $2
done
```
Execute the script with the IP of the kodi instance and 1 sec sleeptime.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

@FernetMenta Your opinion would be appreciated.

FernetMenta commented:
this is wrong. CGraphicContext::SetVideoResolution is supposed to be a synchronous call

a1rwulf commented:
Any ideas on how to fix this deadlock are appreciated.
I can't seem to find a good way how to make sure those 2 calls don't deadlock other then removing the infinite timeout in `StopThread` here:
https://github.com/a1rwulf/xbmc/blob/video-deadlock/xbmc/threads/Thread.cpp#L204

FernetMenta commented:
What lock exactly is causes the deadlock?

FernetMenta commented:
I see. The problem (again) is that the main thread and the render thread are the same. The main thread is blocked by a synchronous call, the stop cmd, but the render thread is required for doing work.
Either split the render and main thread or make the stop cmd async.

a1rwulf commented:
Exactly.
Mainthread's message loop is processing the `TMSG_MEDIA_STOP` message and ultimately calls `CloseFile` which waits for the VideoPlayer-Thread to stop.
However the VideoPlayer-Thread sets the refreshrate with a synchronous call through `AppMessenger` which wants to do the work on the Mainthread as well.

By making the stop cmd async, you mean offloading the `CApplication::StopPlaying` or `CApplicationPlayer::ClosePlayer` into a background thread?

FernetMenta commented:
> By making the stop cmd async, you mean offloading the CApplication::StopPlaying or CApplicationPlayer::ClosePlayer into a background thread?

No, that wouldn't make much sense. I meant making the stop method of players async. Not waiting for the active thread to exit but returning immediately and signalling final stop via an event to CApp. Just like startup is done. That's of course not just a 2-lines change.

a1rwulf commented:
> > By making the stop cmd async, you mean offloading the CApplication::StopPlaying or CApplicationPlayer::ClosePlayer into a background thread?
> 
> No, that wouldn't make much sense. I meant making the stop method of players async. Not waiting for the active thread to exit but returning immediately and signalling final stop via an event to CApp. Just like startup is done. That's of course not just a 2-lines change.

Ah OK, got it.
I'll try to implement that.

a1rwulf commented:
@FernetMenta I tried to figure out what I need to do to make the Stop async.

As a first step I removed the `StopThread` call from CloseFile so the mainthread doesn't get blocked.
Because of the `ResetPlayer` call in https://github.com/xbmc/xbmc/pull/16429/commits/ad0412433017fa42ac6291b6843630f34cc4af0e#diff-50cd1739fc49c125a0f2d2ae53bc932eL35 where `VideoPlayer` gets destroyed this caused an immediate crash.

Hence, I removed that call and wanted to do it later when `VideoPlayer` is really finished.
When I checked the flow I found that this works with the current code already:

The `OnExit` handler of `VideoPlayer` which is called when the VideoPlayer-Thread finishes, calls it's `cb->OnPlayBackStopped();` callback.

This effectively executes `void CApplication::OnPlayBackStopped()` which sends `CGUIMessage msg(GUI_MSG_PLAYBACK_STOPPED, 0, 0);`.

The handler-code of this message calls `PlaybackCleanup();` which will call `m_appPlayer.ResetPlayer();` and thus destroy VideoPlayer and it's workerthreads.

Is this what you meant, or is there more I need to take care of?

a1rwulf commented:
I've tested a few hours now with various playback scripts.
The freeze is gone, and I've seen no crashes so far.

FernetMenta commented:
@a1rwulf sorry to say but every singly line of this change is wrong.  As I mentioned, you can't accomplish this with just two lines and certainly not by: +6 -20.

a1rwulf commented:
@FernetMenta no problem at all. Some guidance would be appreciated, I took quite some time to understand the control flow which I described above.
Making `CloseFile` not wait for the VideoPlayer-Thread to stop seems mandatory to me, to unblock the mainthread.
But I don't know where to go from there.
I took a look at the "Make OpenFile asynchronous"-PR to understand the changes of making Open async, but I couldn't really see how this helps me for this task.
The main problem I have is that immediately after CloseFile, the player gets destroyed which of course can only happen when all related threads are shutdown already.

FernetMenta commented:
In an asynchronous world you have to track states. By removing StopThread you allow the App to open a new file while the last has not finished. This creates races. If Players were not singletons, you could create a new player object for the new file. But resolving this design flaw is even more work.
You could use a jobqueue for stopping player like VP uses it for sending outbound events, i.e. OnAVStarted.  The main thread needs to track state asynchronously. Means CApp must not allow next start until player has singled final stop.

a1rwulf commented:
Thx for clarifying, I think I'm able to trigger the mentioned race condition when using a keyboard.
Seems it doesn't happen when I play through the json Player.Open - I assume because json sends TMSG_MEDIA_PLAY via ApplicationMessenger which already queues the request and makes it slow enough - this might still be "luck" as the GUIMSG_STOP is actually in a different queue.

a1rwulf committed:
VideoPlayer: Correctly handle TMSG_RENDERER_PREINIT/UNINIT

VideoRenderer's PreInit and UnInit need to run on the
mainthread, RenderManager contained code for it, but the
sent message have never been handled.
Also as the real job is done by the mainthread, the early
returns after waiting for the signaled events were missing.

a1rwulf committed:
ApplicationPlayer: Don't reset the player instance when closing

The player reset in here is redundant as the PlaybackCleanup
will take care of it.
This also allows ClosePlayer to be called from a worker thread.

a1rwulf committed:
Application: Async player start/stop

In order to fix a deadlock when users
start/stop very fast, we offload player
close into a background job as part of a
jobqueue.
In order to make sure there are no races
between subsequent opens, we also put it
in the same jobqueue.
That makes sure the playerstop is really
done before the next open can happen.

Meanwhile the app stays responsive to the
user, because nothing blocks the mainthread.

a1rwulf committed:
VideoPlayer: Uninitialize RenderManager in OnExit

As CloseFile can now be called from background
threads, we run into deadlocks between the
background worker and VP, because both want
stuff to be done on the mainthread at cleanup.

Moving the render uninit into the OnExit method
makes sure only the VP thread does cleanup that
needs to be done on the mainthread.

a1rwulf commented:
@FernetMenta would you mind taking a look again?

FernetMenta commented on xbmc/cores/IPlayer.h:
```diff
@@ -208,6 +208,8 @@ class IPlayer
    */
   virtual void Render(bool clear, uint32_t alpha = 255, bool gui = true) {};
   virtual void FlushRenderer() {};
+  virtual void PreInitRenderer() {};
```
pulling implementation details up here is an architectural disaster. not only that CApp should not even know that there is something like a renderer (and other players have none), exposing this is completely unsecure. calling this method at some other place/time will just crash the app.
sorry to say, this is an ugly hack.

a1rwulf replied:
This is the original commit:
https://github.com/xbmc/xbmc/commit/7cd6b19131c54d5f11456ce284d066a3db124daf

I wonder how this was expected to work?
It sends a signal that is never handled, thus running always in a timeout when not called from the mainthread.
Then it doesn't return at all, tries to run the same code in the wrong threadcontext and deadlock or crash.
Why adding a message that is never handled?
For me it looked like handling the message in the mainthread was the missing part...

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp:
```diff
@@ -357,6 +357,12 @@ void CRenderManager::PreInit()
+    }
+    else
+    {
+      // Mainthread did it's work, we're done!
```
guessing

a1rwulf commented:
OK, seems that this leads to nowhere and everything I try is wrong.

Let me try to be more specific in my questions:
* Is the general approach, that is introducing a jobqueue for close/open of the player the kind of synchronization you had in mind at all?
* If so, `VideoPlayer::Open` and `VideoPlayer::Close` will now not run on the mainthread anymore, they will run inside a backgroundworker, hence `RenderManager::PreInit` and `RenderManager::UnInit` cannot be called anymore, cause they seem to need the mainthread.
I don't know how to marshal them onto the mainthread without exposing those methods to upper layers. Any ideas?
* The third issue that I encountered can be discussed later. Let's do that step-by-step.

FernetMenta commented:
> Is the general approach, that is introducing a jobqueue for close/open of the player in CApplication the kind of synchronization you had in mind at all?

Nope because it violates the principle of information hiding.
I meant that the task of stopping could be delegated to some VP internal job, then return immediately in order not to block the main thread. As a result CApp has to deal with the asynchronous behavior, i.e. by introducing states like "player stopping". 
This PR does not consider asynchronous behavior. User would be able to start player while stop is still in process.
If the main thread were not the render thread, it would be easy. But nobody seems to be interested in fixing the root cause. So fighting the symptoms results in a lot of work too.

> FlushRenderer is exposed to CApplication since ages

Correct, this is ancient code. Now VP registers into render thread by calling RegisterRenderLoop. Then render thread calls FrameMove on VP, this calls FrameMove on RenderManager. No need for appMessenger anymore.

a1rwulf commented:
> If the main thread were not the render thread, it would be easy. But nobody seems to be interested in fixing the root cause. So fighting the symptoms results in a lot of work too.

Yeah, honestly I think I'll scratch the async approach and look into splitting the threads.
Looks like that would be way more beneficial in the end, I'll most likely reach out to you in that regard with questions.
